crates/libs/core/src/runtime/stateful.rs

```diff
@@ -111,43 +128,132 @@ pub trait LocalReplicator: Send + Sync + 'static {
+// primary gets the new epoch from change_role() call. Secondary should fence/reject
+// operations from the old primary with an older epoch.
+
+/// TODO: primary replicator has no public documentation, this is gathered unofficially and
+/// is subject to change/correction.
 /// IFabricPrimaryReplicator com interface wrapper.
 #[trait_variant::make(PrimaryReplicator: Send)]
 pub trait LocalPrimaryReplicator: Replicator {
     // SF calls this to indicate that possible data loss has occurred (write quorum loss),
     // returns is isStateChanged. If true, SF will re-create other secondaries.
     // The default SF impl might be a pass through to the state provider.
     async fn on_data_loss(&self, cancellation_token: CancellationToken) -> crate::Result<u8>;
+
+    // Remarks on replicator configuration:
+    // At any time the replicator can have one or two configurations. There is always a current
+    // configuration which represents the set of replicas that are participating in replication
+    // along with the current write quorum. In addition there can be a previous configuration
+    // which represents the set of replicas that were in the previous configuration.
+    // When there is both a current and previous configuration the replicator must ensure that
+    // writes are acknowledeged by a write quroum of both configurations.
+
+    /// Informs the replicator there there is a current configuration and a previous configuration.
+    /// Called on primary to inform the set of active secondary replicas that may
+    /// begin to catchup. Idle secondary replicas are not included here.
+    ///
+    /// The total number of replica marked with must_catchup will not exceed the write quorum.
+    /// Secondary to be promoted to new primary is guaranteed to have must_catchup set at some point,
+    /// i.e. it must catch up (have all the data) to be promoted to new primary.
+    ///
+    /// ReplicaInformation:
+    /// current_progress -> The LSN of the replica. -1 if the replicator is already aware of the replica
+    /// (it is in configuration or has been built) otherwise it will be the progress of the remote replica.
+    /// catch_up_capability -> The first LSN of the replica. Similar to current_progress.
+    /// must_catchup -> Set to true only for one replica in the current configuration.
     fn update_catch_up_replica_set_configuration(
         &self,
         currentconfiguration: &ReplicaSetConfig,
         previousconfiguration: &ReplicaSetConfig,
     ) -> crate::Result<()>;
+
+    /// Informs the replicator about the current replica set configuration, and there
+    /// is no longer a previous configuration.
+    /// Remarks:
+    /// Replicas here are not marked as must_catchup.
+    fn update_current_replica_set_configuration(
+        &self,
+        currentconfiguration: &ReplicaSetConfig,
+    ) -> crate::Result<()>;
+
+    /// Called on primary to wait for the quorum of replicas to catch up, before
```

Comment:
Might be worth defining what "catch up" means from the SF PoV

Reply:
Added in the remarks.